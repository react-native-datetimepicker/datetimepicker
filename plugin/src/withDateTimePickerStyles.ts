import {
  AndroidConfig,
  ConfigPlugin,
  withAndroidColors,
  withAndroidColorsNight,
  withAndroidStyles,
} from '@expo/config-plugins';

type ResourceXML = AndroidConfig.Resources.ResourceXML;

const moduleName = 'RNDatetimePicker: ';

const DATE_PICKER_ALLOWED_ATTRIBUTES = {
  colorAccent: {attrName: 'colorAccent'},
  colorControlActivated: {
    attrName: 'colorControlActivated',
  },
  colorControlHighlight: {
    attrName: 'colorControlHighlight',
  },
  selectableItemBackgroundBorderless: {
    attrName: 'android:selectableItemBackgroundBorderless',
  },
  textColor: {
    attrName: 'android:textColor',
  },
  textColorPrimary: {
    attrName: 'android:textColorPrimary',
  },
  textColorPrimaryInverse: {
    attrName: 'android:textColorPrimaryInverse',
  },
  textColorSecondary: {
    attrName: 'android:textColorSecondary',
  },
  textColorSecondaryInverse: {
    attrName: 'android:textColorSecondaryInverse',
  },
  windowBackground: {
    attrName: 'android:windowBackground',
  },
} as const;

type ColorDefinition = {
  light: string;
  dark?: string;
};
type DatePickerProps = {
  [key in keyof typeof DATE_PICKER_ALLOWED_ATTRIBUTES]?: ColorDefinition;
};

const TIME_PICKER_ALLOWED_ATTRIBUTES = {
  background: {
    attrName: 'android:background',
  },
  headerBackground: {
    attrName: 'android:headerBackground',
  },
  numbersBackgroundColor: {
    attrName: 'android:numbersBackgroundColor',
  },
  numbersSelectorColor: {
    attrName: 'android:numbersSelectorColor',
  },
  numbersTextColor: {
    attrName: 'android:numbersTextColor',
  },
} as const;

type TimePickerProps = {
  [key in keyof typeof TIME_PICKER_ALLOWED_ATTRIBUTES]?: ColorDefinition;
};

type Options = {
  android?: {
    datePicker?: DatePickerProps;
    timePicker?: TimePickerProps;
  };
};

const DATE_PICKER_THEME_ATTRIBUTE = 'android:datePickerDialogTheme';
const DATE_PICKER_STYLE_NAME = 'DatePickerDialogTheme';

const TIME_PICKER_THEME_ATTRIBUTE = 'android:timePickerStyle';
const TIME_PICKER_STYLE_NAME = 'TimePickerTheme';

const {assignStylesValue, getAppThemeLightNoActionBarGroup} =
  AndroidConfig.Styles;

const {assignColorValue} = AndroidConfig.Colors;

const insertColorEntries = (
  android: Options['android'] = {},
  config: {modResults: ResourceXML},
  themedColorExtractor: (
    color: ColorDefinition,
    attrName: string,
  ) => string | undefined,
) => {
  for (const {theme, attrPrefix} of [
    {theme: android.datePicker, attrPrefix: 'datePicker'},
    {theme: android.timePicker, attrPrefix: 'timePicker'},
  ]) {
    config.modResults = setAndroidColors(
      config.modResults,
      attrPrefix,
      themedColorExtractor,
      theme,
    );
  }
};

const setAndroidColors = (
  colors: ResourceXML,
  attrPrefix: string,
  themedColorExtractor: (
    color: ColorDefinition,
    attrName: string,
  ) => string | undefined,
  theme?: DatePickerProps | TimePickerProps,
): ResourceXML => {
  if (!theme) {
    return colors;
  }
  colors = Object.entries(theme).reduce((acc, [attrName, colors]) => {
    const color = {
      name: `${attrPrefix}_${attrName}`,
      value: themedColorExtractor(colors, attrName) ?? null,
    };
    return assignColorValue(acc, color);
  }, colors);

  return colors;
};

const withDateTimePickerStyles: ConfigPlugin<Options> = (
  baseConfig,
  options = {},
) => {
  const {android = {}} = options;

  let newConfig = withAndroidColors(baseConfig, (config) => {
    insertColorEntries(android, config, (color, attrName) => {
      const value = color.light;
      if (!value) {
        throw new Error(
          `${moduleName}A light color value was not provided for "${attrName}". Providing at least a light color is required.`,
        );
      }
      return value;
    });
    return config;
  });

  newConfig = withAndroidColorsNight(newConfig, (config) => {
    insertColorEntries(android, config, (color) => color.dark);
    return config;
  });

  newConfig = withAndroidStyles(newConfig, (config) => {
    config.modResults = setAndroidPickerStyles(
      config.modResults,
      android.datePicker,
      {
        styleName: DATE_PICKER_STYLE_NAME,
        parentStyle: 'Theme.AppCompat.Light.Dialog',
        themeAttribute: DATE_PICKER_THEME_ATTRIBUTE,
        allowedAttributes: DATE_PICKER_ALLOWED_ATTRIBUTES,
        attrPrefix: 'datePicker',
      },
    );

    config.modResults = setAndroidPickerStyles(
      config.modResults,
      android.timePicker,
      {
        styleName: TIME_PICKER_STYLE_NAME,
        parentStyle: 'android:Widget.Material.Light.TimePicker',
        themeAttribute: TIME_PICKER_THEME_ATTRIBUTE,
        allowedAttributes: TIME_PICKER_ALLOWED_ATTRIBUTES,
        attrPrefix: 'timePicker',
      },
    );

    return config;
  });

  return newConfig;
};

const setAndroidPickerStyles = (
  styles: ResourceXML,
  theme: DatePickerProps | TimePickerProps | undefined,
  config: {
    styleName: string;
    parentStyle: string;
    themeAttribute: string;
    allowedAttributes:
      | typeof DATE_PICKER_ALLOWED_ATTRIBUTES
      | typeof TIME_PICKER_ALLOWED_ATTRIBUTES;
    attrPrefix: string;
  },
): ResourceXML => {
  if (!theme) {
    return styles;
  }
  const {
    allowedAttributes,
    styleName,
    parentStyle,
    themeAttribute,
    attrPrefix,
  } = config;

  styles = Object.keys(theme).reduce((acc, userFacingAttrName) => {
    const entry =
      allowedAttributes[
        userFacingAttrName as keyof (DatePickerProps | TimePickerProps)
      ];
    if (!entry) {
      throw new Error(
        `${moduleName}Invalid attribute name: ${userFacingAttrName}. Supported are ${Object.keys(
          allowedAttributes,
        ).join(', ')}`,
      );
    }
    const {attrName} = entry;

    return assignStylesValue(acc, {
      add: true,
      parent: {
        name: styleName,
        parent: parentStyle,
      },
      name: attrName,
      value: `@color/${attrPrefix}_${userFacingAttrName}`,
    });
  }, styles);

  styles = assignStylesValue(styles, {
    add: true,
    parent: getAppThemeLightNoActionBarGroup(),
    name: themeAttribute,
    value: `@style/${styleName}`,
  });

  return styles;
};

export default withDateTimePickerStyles;
