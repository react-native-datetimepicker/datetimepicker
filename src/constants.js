/**
 * @format
 * @flow strict-local
 */
export const MIN_MS = 60000;

export const ANDROID_DISPLAY = Object.freeze({
  default: 'default',
  spinner: 'spinner',
  clock: 'clock',
  calendar: 'calendar',
});

export const EVENT_TYPE_SET = 'set';
export const EVENT_TYPE_DISMISSED = 'dismissed';
export const ANDROID_EVT_TYPE = Object.freeze({
  set: EVENT_TYPE_SET,
  dismissed: EVENT_TYPE_DISMISSED,
  neutralButtonPressed: 'neutralButtonPressed',
});

export const IOS_DISPLAY = Object.freeze({
  default: 'default',
  spinner: 'spinner',
  compact: 'compact',
  inline: 'inline',
});

const COMMON_MODES = Object.freeze({
  date: 'date',
  time: 'time',
});

export const ANDROID_MODE = COMMON_MODES;

export const WINDOWS_MODE = COMMON_MODES;

export const IOS_MODE = Object.freeze({
  ...COMMON_MODES,
  datetime: 'datetime',
  countdown: 'countdown',
});

export const DAY_OF_WEEK = Object.freeze({
  Sunday: 0,
  Monday: 1,
  Tuesday: 2,
  Wednesday: 3,
  Thursday: 4,
  Friday: 5,
  Saturday: 6,
});

export const DATE_SET_ACTION = 'dateSetAction';
export const TIME_SET_ACTION = 'timeSetAction';
export const DISMISS_ACTION = 'dismissedAction';

export const NEUTRAL_BUTTON_LABEL = 'neutralButtonLabel';
export const NEUTRAL_BUTTON_ACTION = 'neutralButtonAction';
