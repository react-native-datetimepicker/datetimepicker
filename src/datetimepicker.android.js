/**
 * @format
 * @flow strict-local
 */
import {
  DATE_SET_ACTION,
  TIME_SET_ACTION,
  DISMISS_ACTION,
  NEUTRAL_BUTTON_ACTION,
  ANDROID_DISPLAY,
  ANDROID_MODE,
  MIN_MS,
} from './constants';
import pickers from './picker';
import invariant from 'invariant';
import {useEffect} from 'react';

import type {AndroidEvent, AndroidNativeProps} from './types';

function validateProps(props: AndroidNativeProps) {
  const {mode, value, display} = props;
  invariant(value, 'A date or time should be specified as `value`.');
  invariant(
    !(display === ANDROID_DISPLAY.calendar && mode === ANDROID_MODE.time) &&
      !(display === ANDROID_DISPLAY.clock && mode === ANDROID_MODE.date),
    `display: ${display} and mode: ${mode} cannot be used together.`,
  );
}

function getPicker({
  mode,
  value,
  display,
  is24Hour,
  minimumDate,
  maximumDate,
  neutralButtonLabel,
  minuteInterval,
  timeZoneOffsetInMinutes,
}) {
  switch (mode) {
    case ANDROID_MODE.time:
      return pickers[mode].open({
        value,
        display,
        minuteInterval,
        is24Hour,
        neutralButtonLabel,
        timeZoneOffsetInMinutes,
      });
    default:
      return pickers[ANDROID_MODE.date].open({
        value,
        display,
        minimumDate,
        maximumDate,
        neutralButtonLabel,
        timeZoneOffsetInMinutes,
      });
  }
}

function timeZoneOffsetDateSetter(date, timeZoneOffsetInMinutes) {
  if (typeof timeZoneOffsetInMinutes === 'number') {
    // FIXME this causes a bug. repro: set tz offset to zero, and then keep opening and closing the calendar picker
    // https://github.com/react-native-datetimepicker/datetimepicker/issues/528
    const offset = date.getTimezoneOffset() + timeZoneOffsetInMinutes;
    const shiftedDate = new Date(date.getTime() - offset * MIN_MS);
    return shiftedDate;
  }
  return date;
}

export default function RNDateTimePicker(props: AndroidNativeProps) {
  validateProps(props);
  const {
    mode = ANDROID_MODE.date,
    display = ANDROID_DISPLAY.default,
    value,
    onChange,
    is24Hour,
    minimumDate,
    maximumDate,
    neutralButtonLabel,
    minuteInterval,
    timeZoneOffsetInMinutes,
  } = props;
  const valueTimestamp = value.getTime();

  useEffect(() => {
    // This effect runs on unmount / with mode change, and will ensure the picker is closed.
    // This allows for controlling the opening state of the picker through declarative logic in jsx.
    return () => (pickers[mode] ?? pickers[ANDROID_MODE.date]).dismiss();
  }, [mode]);

  useEffect(
    function showOrUpdatePicker() {
      const picker = getPicker({
        mode,
        value: valueTimestamp,
        display,
        is24Hour,
        minimumDate,
        maximumDate,
        neutralButtonLabel,
        minuteInterval,
        timeZoneOffsetInMinutes,
      });

      picker.then(
        function resolve({action, day, month, year, minute, hour}) {
          let date = new Date(valueTimestamp);
          const event: AndroidEvent = {
            type: 'set',
            nativeEvent: {},
          };

          switch (action) {
            case DATE_SET_ACTION:
              date.setFullYear(year, month, day);
              date = timeZoneOffsetDateSetter(date, timeZoneOffsetInMinutes);
              event.nativeEvent.timestamp = date;
              onChange(event, date);
              break;

            case TIME_SET_ACTION:
              date.setHours(hour, minute);
              date = timeZoneOffsetDateSetter(date, timeZoneOffsetInMinutes);
              event.nativeEvent.timestamp = date;
              onChange(event, date);
              break;

            case NEUTRAL_BUTTON_ACTION:
              event.type = 'neutralButtonPressed';
              onChange(event);
              break;

            case DISMISS_ACTION:
            default:
              event.type = 'dismissed';
              onChange(event);
              break;
          }
        },
        function reject(error) {
          // ignore or throw `activity == null` error
          throw error;
        },
      );
    },
    // the android dialog, when presented, will actually ignore updates to all props other than `value`
    // we need to change the behavior as described in https://github.com/react-native-datetimepicker/datetimepicker/pull/327#issuecomment-723160992
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [onChange, valueTimestamp, mode],
  );

  return null;
}
