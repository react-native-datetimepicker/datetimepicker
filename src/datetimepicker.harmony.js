/**
 * MIT License
 *
 * Copyright (C) 2023 Huawei Device Co., Ltd.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import RNDateTimePicker from './picker';
import {dateToMilliseconds, sharedPropsValidation} from './utils';
import {
  IOS_DISPLAY,
  EVENT_TYPE_SET,
  EVENT_TYPE_DISMISSED,
  IOS_MODE,
} from './constants';
import * as React from 'react';
//import {Platform} from 'react-native';

import type {
  DateTimePickerEvent,
  NativeEventIOS,
  IOSNativeProps,
  IOSDisplay,
} from './types';

// const getDisplaySafe = (display: IOSDisplay): IOSDisplay=>{
//     const majorVersionIOS = parseInt(Platfrom.Version, 10);
//     if(display === IOS_DISPLAY.inline && majorVersionIOS < 14){
//         // inline is available since 14.0
//         return IOS_DISPLAY.spinner;
//     }
//     if(majorVersionIOS < 14){
//         // NOTE this should compare against 13.4, not 14 according to https://developer.apple.com/documentation/uikit/uidatepickerstyle/uidatepickerstylecompact?changes=latest_minor&language=objc
//         // but UIDatePickerStyleCompact does not seem to work prior to 14
//         // only the spinner display (UIDatePickerStyleWheels) is thus available below 14
//         return IOS_DISPLAY.spinner;
//     }
//
//     return display
// }

export default function Picker({
  value,
  locale,
  maximumDate,
  minimumDate,
  minuteInterval,
  timeZoneOffsetInMinutes,
  timeZoneName,
  textColor,
  accentColor,
  themeVariant,
  onChange,
  mode = IOS_MODE.date,
  display: providedDisplay = IOS_DISPLAY.default,
  disabled = false,
  is24Hour = true,
  ...other
}: IOSNativeProps): React.Node {
  sharedPropsValidation({value, timeZoneOffsetInMinutes, timeZoneName});

  // const display = getDisplaySafe(providedDisplay)
  const display = providedDisplay;

  const _onChange = (event: NativeEventIOS) => {
    const timestamp = event.nativeEvent.timestamp;
    const unifiedEvent: DateTimePickerEvent = {
      ...event,
      type: EVENT_TYPE_SET,
    };

    const date = timestamp !== undefined ? new Date(timestamp) : undefined;

    onChange && onChange(unifiedEvent, date);
  };

  const onDismiss = () => {
    // TODO introduce separate onDismissed event listener
    onChange &&
      onChange(
        {
          type: EVENT_TYPE_DISMISSED,
          nativeEvent: {
            timestamp: value.getTime(),
            utcOffset: 0, // TODO vonovak - the dismiss event should not carry any date information
          },
        },
        value,
      );
  };

  return (
    <RNDateTimePicker
      {...other}
      date={dateToMilliseconds(value)}
      locale={locale !== null && locale !== '' ? locale : undefined}
      maximumDate={dateToMilliseconds(maximumDate)}
      minimumDate={dateToMilliseconds(minimumDate)}
      mode={mode}
      minuteInterval={minuteInterval}
      timeZoneOffsetInMinutes={timeZoneOffsetInMinutes}
      timeZoneName={timeZoneName}
      onChange={_onChange}
      onPickerDismiss={onDismiss}
      textColor={textColor}
      accentColor={accentColor}
      themeVariant={themeVariant}
      onStartShouldSetResponder={() => true}
      onResponderTerminationRequest={() => false}
      displayIOS={display}
      enabled={disabled !== true}
      is24Hour={is24Hour}
    />
  );
}
