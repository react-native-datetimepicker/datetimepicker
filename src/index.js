/**
 * @format
 * @flow strict-local
 */
import RNDateTimePicker from './datetimepicker';
export * from './eventCreators';
export {DateTimePickerAndroid} from './DateTimePickerAndroid';

export default RNDateTimePicker;
