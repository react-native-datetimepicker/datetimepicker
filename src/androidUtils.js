/**
 * @format
 * @flow strict-local
 */
import {ANDROID_DISPLAY, ANDROID_MODE, MIN_MS} from './constants';
import pickers from './picker';
import type {AndroidNativeProps, DateTimePickerResult} from './types';
import {sharedPropsValidation} from './utils';
import invariant from 'invariant';
import {processColor} from 'react-native';

type Timestamp = number;

type ProcessedButton = {
  title: string,
  textColor: $Call<typeof processColor>,
};

type Params = {
  value: Timestamp,
  display: AndroidNativeProps['display'],
  is24Hour: AndroidNativeProps['is24Hour'],
  minimumDate: AndroidNativeProps['minimumDate'],
  maximumDate: AndroidNativeProps['maximumDate'],
  minuteInterval: AndroidNativeProps['minuteInterval'],
  timeZoneOffsetInMinutes: AndroidNativeProps['timeZoneOffsetInMinutes'],
  dialogButtons: {
    positive: ProcessedButton,
    negative: ProcessedButton,
    neutral: ProcessedButton,
  },
};

export type PresentPickerCallback = (Params) => Promise<DateTimePickerResult>;

function getOpenPicker(
  mode: AndroidNativeProps['mode'],
): PresentPickerCallback {
  switch (mode) {
    case ANDROID_MODE.time:
      return ({
        value,
        display,
        is24Hour,
        minuteInterval,
        timeZoneOffsetInMinutes,
        dialogButtons,
      }: Params) =>
        // $FlowFixMe - `AbstractComponent` [1] is not an instance type.
        pickers[mode].open({
          value,
          display,
          minuteInterval,
          is24Hour,
          timeZoneOffsetInMinutes,
          dialogButtons,
        });
    default:
      return ({
        value,
        display,
        minimumDate,
        maximumDate,
        timeZoneOffsetInMinutes,
        dialogButtons,
      }: Params) =>
        // $FlowFixMe - `AbstractComponent` [1] is not an instance type.
        pickers[ANDROID_MODE.date].open({
          value,
          display,
          minimumDate,
          maximumDate,
          timeZoneOffsetInMinutes,
          dialogButtons,
        });
  }
}

function timeZoneOffsetDateSetter(
  date: Date,
  timeZoneOffsetInMinutes: ?number,
): Date {
  if (typeof timeZoneOffsetInMinutes === 'number') {
    // FIXME this causes a bug. repro: set tz offset to zero, and then keep opening and closing the calendar picker
    // https://github.com/react-native-datetimepicker/datetimepicker/issues/528
    const offset = date.getTimezoneOffset() + timeZoneOffsetInMinutes;
    const shiftedDate = new Date(date.getTime() - offset * MIN_MS);
    return shiftedDate;
  }
  return date;
}

function validateAndroidProps(props: AndroidNativeProps) {
  sharedPropsValidation({value: props?.value});
  const {mode, display} = props;
  invariant(
    !(display === ANDROID_DISPLAY.calendar && mode === ANDROID_MODE.time) &&
      !(display === ANDROID_DISPLAY.clock && mode === ANDROID_MODE.date),
    `display: ${display} and mode: ${mode} cannot be used together.`,
  );
  if (
    props?.positiveButtonLabel !== undefined ||
    props?.negativeButtonLabel !== undefined ||
    props?.neutralButtonLabel !== undefined
  ) {
    console.warn(
      'positiveButtonLabel, negativeButtonLabel and neutralButtonLabel are deprecated.' +
        'Use positive / negative / neutralButton prop instead.',
    );
  }
}
export {getOpenPicker, timeZoneOffsetDateSetter, validateAndroidProps};
