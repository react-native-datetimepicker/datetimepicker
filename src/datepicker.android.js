/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 */
import {DATE_SET_ACTION, DISMISS_ACTION, ANDROID_DISPLAY} from './constants';
import {toMilliseconds} from './utils';
import RNDatePickerAndroid from './specs/NativeModuleDatePicker';
import type {DatePickerOptions, DateTimePickerResult} from './types';

export default class DatePickerAndroid {
  /**
   * Opens the standard Android date picker dialog.
   *
   * The available keys for the `options` object are:
   *
   *   - `value` (`Date` object) - date to show by default
   *   - `minimumDate` (`Date` object) - minimum date that can be selected
   *   - `maximumDate` (`Date` object) - maximum date that can be selected
   *   - `display` (`enum('calendar', 'spinner', 'default')`) - To set the date-picker display to calendar/spinner/default
   *     - 'calendar': Show a date picker in calendar mode.
   *     - 'spinner': Show a date picker in spinner mode.
   *     - 'default': Show a default native date picker(spinner/calendar) based on android versions.
   *
   * Returns a Promise which will be invoked an object containing `action`, `year`, `month` (0-11),
   * `day` if the user picked a date. If the user dismissed the dialog, the Promise will
   * still be resolved with action being `DatePickerAndroid.dismissedAction` and all the other keys
   * being undefined. **Always** check whether the `action` before reading the values.
   *
   * Note the native date picker dialog has some UI glitches on Android 4 and lower
   * when using the `minimumDate` and `maximumDate` options.
   */
  static async open(options: DatePickerOptions): Promise<DateTimePickerResult> {
    toMilliseconds(options, 'value', 'minimumDate', 'maximumDate');
    options.display = options.display || ANDROID_DISPLAY.default;

    return RNDatePickerAndroid.open(options);
  }

  static async dismiss(): Promise<boolean> {
    return RNDatePickerAndroid.dismiss();
  }

  /**
   * A date has been selected.
   */
  static +dateSetAction: 'dateSetAction' = DATE_SET_ACTION;
  /**
   * The dialog has been dismissed.
   */
  static +dismissedAction: 'dismissedAction' = DISMISS_ACTION;
}
