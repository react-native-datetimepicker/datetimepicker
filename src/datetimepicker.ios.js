/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * This is a controlled component version of RNDateTimePicker
 *
 * @format
 * @flow strict-local
 */
import RNDateTimePicker from './picker';
import {sharedPropsValidation, toMilliseconds} from './utils';
import {IOS_DISPLAY, ANDROID_MODE, EVENT_TYPE_SET} from './constants';
import invariant from 'invariant';
import * as React from 'react';
import {getPickerHeightStyle} from './layoutUtilsIOS';
import {Platform, StyleSheet} from 'react-native';

import type {
  NativeEventIOS,
  NativeRef,
  IOSNativeProps,
  DatePickerOptions,
  IOSDisplay,
} from './types';
import type {DateTimePickerEvent} from './types';

const getDisplaySafe = (display: IOSDisplay): IOSDisplay => {
  const majorVersionIOS = parseInt(Platform.Version, 10);
  if (display === IOS_DISPLAY.inline && majorVersionIOS < 14) {
    // inline is available since 14.0
    return IOS_DISPLAY.spinner;
  }
  if (majorVersionIOS < 14) {
    // NOTE this should compare against 13.4, not 14 according to https://developer.apple.com/documentation/uikit/uidatepickerstyle/uidatepickerstylecompact?changes=latest_minor&language=objc
    // but UIDatePickerStyleCompact does not seem to work prior to 14
    // only the spinner display (UIDatePickerStyleWheels) is thus available below 14
    return IOS_DISPLAY.spinner;
  }

  return display;
};

export default function Picker({
  value,
  locale,
  maximumDate,
  minimumDate,
  style,
  testID,
  minuteInterval,
  timeZoneOffsetInMinutes,
  textColor,
  accentColor,
  themeVariant,
  onChange,
  mode = ANDROID_MODE.date,
  display: providedDisplay = IOS_DISPLAY.default,
  disabled = false,
}: IOSNativeProps): React.Node {
  sharedPropsValidation({value});

  const [heightStyle, setHeightStyle] = React.useState(undefined);
  const _picker: NativeRef = React.useRef(null);
  const display = getDisplaySafe(providedDisplay);

  React.useEffect(
    function ensureNativeIsInSyncWithJS() {
      const {current} = _picker;

      if (value && onChange && current) {
        const timestamp = value.getTime();
        // $FlowFixMe Cannot call `current.setNativeProps` because property `setNativeProps` is missing in `AbstractComponent` [1].
        current.setNativeProps({
          date: timestamp,
        });
      }
    },
    [onChange, value],
  );

  React.useEffect(
    function ensureCorrectHeight() {
      const height = getPickerHeightStyle(display, mode);
      if (height instanceof Promise) {
        height.then(setHeightStyle);
      } else {
        setHeightStyle(height);
      }
    },
    [display, mode],
  );

  const _onChange = (event: NativeEventIOS) => {
    const timestamp = event.nativeEvent.timestamp;
    // $FlowFixMe Cannot assign object literal to `unifiedEvent` because number [1] is incompatible with undefined [2] in property `nativeEvent.timestamp`.
    const unifiedEvent: DateTimePickerEvent = {...event, type: EVENT_TYPE_SET};

    const date = timestamp !== undefined ? new Date(timestamp) : undefined;

    onChange && onChange(unifiedEvent, date);
  };

  invariant(value, 'A date or time should be specified as `value`.');

  if (!heightStyle) {
    // wait for height to be available in state
    return null;
  }

  const dates: DatePickerOptions = {value, maximumDate, minimumDate};
  toMilliseconds(dates, 'value', 'minimumDate', 'maximumDate');

  return (
    // $FlowFixMe - dozen of flow errors
    <RNDateTimePicker
      testID={testID}
      ref={_picker}
      style={StyleSheet.compose(heightStyle, style)}
      date={dates.value}
      locale={locale !== null && locale !== '' ? locale : undefined}
      maximumDate={dates.maximumDate}
      minimumDate={dates.minimumDate}
      mode={mode}
      minuteInterval={minuteInterval}
      timeZoneOffsetInMinutes={timeZoneOffsetInMinutes}
      onChange={_onChange}
      textColor={textColor}
      accentColor={accentColor}
      themeVariant={themeVariant}
      onStartShouldSetResponder={() => true}
      onResponderTerminationRequest={() => false}
      displayIOS={display}
      enabled={disabled !== true}
    />
  );
}
