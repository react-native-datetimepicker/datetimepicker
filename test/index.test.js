import React from 'react';
import DatePicker from '../src/index.js';
import RNDateTimePickerIOS from '../src/picker.ios';
import AndroidDateTimePicker from '../src/datetimepicker.android';
import {render, fireEvent, waitFor} from '@testing-library/react-native';
import {EVENT_TYPE_SET} from '../src/constants';

const DATE = 1376949600000;

const renderPicker = async (props) => {
  const utils = render(<DatePicker value={new Date(DATE)} {...props} />);
  await waitFor(() => utils.UNSAFE_getByType(DatePicker));
  return utils;
};

describe('DatePicker', () => {
  it('renders a native Component', async () => {
    const {toJSON} = await renderPicker();

    expect(toJSON()).toMatchSnapshot();
  });

  it('will pass timestamp to native Component', async () => {
    const date = new Date(156e10);
    const {toJSON} = await renderPicker({
      value: date,
    });

    expect(toJSON()).toHaveProperty('props.date', date.getTime());
  });

  it('calls onChange callback', async () => {
    expect.assertions(4);
    const date = new Date(156e10);

    function onChange(event, dateArg) {
      expect(event).toHaveProperty('type', EVENT_TYPE_SET);
      expect(event).toHaveProperty('nativeEvent');
      expect(event.nativeEvent).toHaveProperty('timestamp', date.getTime());
      expect(dateArg).toEqual(date);
    }
    const {UNSAFE_getByType} = await renderPicker({onChange});

    fireEvent(UNSAFE_getByType(RNDateTimePickerIOS), 'onChange', {
      nativeEvent: {
        timestamp: date.getTime(),
      },
    });
  });

  it.each([['time'], ['datetime'], ['countdown']])(
    'renders with mode %s',
    async (mode) => {
      const {toJSON} = await renderPicker({
        mode,
      });
      expect(toJSON()).toMatchSnapshot();
    },
  );

  it.each([
    [{value: 'bogus'}, '`value` prop must be an instance of Date object'],
    [{}, 'A date or time must be specified as `value` prop'],
    [
      {display: 'calendar', mode: 'time', value: new Date()},
      'display: calendar and mode: time cannot be used together.',
    ],
    [
      {display: 'clock', mode: 'date', value: new Date()},
      'display: clock and mode: date cannot be used together.',
    ],
  ])(
    'AndroidDateTimePicker throws when invalid props are passed',
    (props, expected) => {
      expect(() => AndroidDateTimePicker(props)).toThrow(expected);
    },
  );

  it('applies styling to DatePicker', async () => {
    const style = {backgroundColor: 'red'};
    const {toJSON} = await renderPicker({
      style,
    });
    const snapshot = toJSON();
    expect(snapshot).toHaveProperty('props.style', [
      {height: 216},
      {backgroundColor: 'red'},
    ]);
    expect(snapshot).toMatchSnapshot();
  });
});
