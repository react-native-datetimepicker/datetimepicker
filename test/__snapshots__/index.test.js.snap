// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker applies styling to DatePicker 1`] = `
<RNDateTimePicker
  date={1376949600000}
  displayIOS="default"
  enabled={true}
  mode="date"
  onChange={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "height": 216,
      },
      Object {
        "backgroundColor": "red",
      },
    ]
  }
/>
`;

exports[`DatePicker renders a native Component 1`] = `
<RNDateTimePicker
  date={1376949600000}
  displayIOS="default"
  enabled={true}
  mode="date"
  onChange={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "height": 216,
    }
  }
/>
`;

exports[`DatePicker renders with mode countdown 1`] = `
<RNDateTimePicker
  date={1376949600000}
  displayIOS="default"
  enabled={true}
  mode="countdown"
  onChange={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "height": 216,
    }
  }
/>
`;

exports[`DatePicker renders with mode datetime 1`] = `
<RNDateTimePicker
  date={1376949600000}
  displayIOS="default"
  enabled={true}
  mode="datetime"
  onChange={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "height": 216,
    }
  }
/>
`;

exports[`DatePicker renders with mode time 1`] = `
<RNDateTimePicker
  date={1376949600000}
  displayIOS="default"
  enabled={true}
  mode="time"
  onChange={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "height": 216,
    }
  }
/>
`;
